### ISSUE TITLE
Stop gradient on `residual = residual - z_q_i.detach()`

### ISSUE THREAD

**Issue by ds-hwang** [descriptinc/descript-audio-codec#107]
DAC doesn't stop gradient for `residual = residual - z_q_i.detach()`.
https://github.com/descriptinc/descript-audio-codec/blob/c7cfc5d2647e26471dc394f95846a0830e7bec34/dac/nn/quantize.py#L186

However, vector_quantize_pytorch stop gradient on it.
https://github.com/lucidrains/vector-quantize-pytorch/blob/976335f03b259536a06ed88a7adb7248cdb3d17c/vector_quantize_pytorch/residual_vq.py#L386

RVQ is trained greedily, stage by stage: Each quantizer should only be responsible for correcting the residual from the previous quantizers, without those previous quantizers being updated again.

If you don’t stop the gradient: Earlier quantizers will receive gradients based on the later quantizer errors.
That breaks the stage-wise assumption, and learning becomes entangled.
This can destabilise training and lead to inefficient or overlapping codebooks.

What do you think?

**Comment by Alexey322** [comment]
@ds-hwang Did you find out the answer?

**Comment by listener17** [comment]
@ds-hwang: I'm not from Descript, but I feel you made a good catch! 

Also, shouldn't it be: 
residual = residual - z_q_i * mask[:, None, None]
??

**Comment by lixinghe1999** [comment]
from my experience, the performance doesn't change.
Maybe the speech can be faster with detach()
